File: ing/1992.rb
# Why runtime error at 100%?
# 1992.py passes

def quadtree(size, arr, i, j)
  val = arr[i][j]
  same = true

  (0...size).each do |x|
    (0...size).each do |y|
      if arr[i + x][j + y] != val
        same = false
        break
      end
    end
  end

  return val if same

  half = size / 2
  "(#{quadtree(half, arr, i, j) +
    quadtree(half, arr, i, j + half) +
    quadtree(half, arr, i + half, j) +
    quadtree(half, arr, i + half, j + half)})"
end

if __FILE__ == $0
  size = gets.to_i
  arr = []
  size.times do
    arr << gets.chomp!.chars
  end
  puts quadtree(size, arr, 0, 0)
end

File: boj/1992.py
def main():
    N = int(input())
    arr = []
    for _ in range(N):
        arr.append(list(input()))
    print(quadtree(N, arr, 0, 0))


def quadtree(size, arr, i, j):
    a = arr[i][j]

    for x in range(size):
        for y in range(size):
            if arr[i + x][j + y] != a:
                half = size // 2
                return '(' + ''.join([
                    quadtree(half, arr, i, j),
                    quadtree(half, arr, i, j + half),
                    quadtree(half, arr, i + half, j),
                    quadtree(half, arr, i + half, j + half)
                ]) + ')'

    return a


if __name__ == "__main__":
    main()
